### PR TITLE
Bind inherited relational fields, stop four leaks

### LINKED ISSUE #680: The license-year walk spends 90% of its time in --find-copies-harder, and 24 files depend on it
Found while closing [#674](https://github.com/Ambiguous-Interactive/unity-helpers/issues/674), which replaced one `git log --follow` per file with a single history walk in `scripts/license-year-lib.sh`. That walk is now the whole cost, and one flag is nearly all of it.

## The measurement

Over 2,163 tracked `.cs` files in this checkout:

| Subject | Wall clock |
| --- | --- |
| `update-license-headers.sh --dry-run`, before #674 | 9 m 49 s |
| after #674 | **1 m 53 s** |
| the single walk alone, inside that | 1 m 42 s |
| the identical walk **without** `--find-copies-harder` | **1.0 s** |

So the flag costs about 100x, and it is ~90% of what the fixer now takes.

## It cannot simply be dropped

Folding both variants and diffing over all 2,163 paths: **24 files resolve to a different year without it.** One example, so the shape is concrete rather than a count:

```text
Runtime/Core/Serialization/JsonConverters/ColorConverter.cs   2023 -> 2025
```

Today's committed headers agree with the `--find-copies-harder` answer, so removing the flag would make `audit-license-years.sh` report 24 mismatches against files nobody has touched. The flag is load-bearing, not decoration.

## Why it might be cheaper than it looks

`user` time is only about 14 s against 1 m 42 s wall. The rest is I/O, and this container's `.git` is a **bind mount from the Windows host** (`devcontainer.json` `mounts`), which is the slowest possible place to do a rename/copy-detecting full-history walk. A native Linux CI runner has no such penalty, so the number that matters for CI may be much smaller than the number measured here.

**Measure it on a CI runner before optimizing anything.** Optimizing a bind-mount artifact would be work spent on a cost the real gate does not pay — the same trap `.llm/context.md` records for benchmarks that measure the instrument.

## Options, if the CI number is also bad

1. **Persist the folded map**, not just the per-path answers. `.git/license-year-cache` is keyed by path and is the audit's; a walk-result cache is a different artifact with different invalidation (it is valid until `HEAD` moves). #668's rule still binds: a staged-rename answer must never be written to a path-keyed cache that outlives the index.
2. **Narrow the walk's pathspec** to the trees the callers actually ask about.
3. **Split the flag out**: run the cheap walk always, and the expensive one only for the paths the cheap walk did not answer. Whether the 24 differing files are identifiable that way is the thing to check first.

## Acceptance criteria

- The walk's cost is measured **on a CI runner**, not only in a bind-mounted container, and the number is recorded.
- If it is optimized, `audit-license-years.sh --summary` still reports zero mismatches over the real repository, and the 24 files above are named in the test that protects them.
- If it is left alone, that is a recorded decision with the CI number beside it.

Related: #674, #668.

### LINKED ISSUE #687: No check project compiles Runtime/Integrations, so no WUH rule has ever run there
`Runtime/Integrations/**` — 19 files across Reflex, VContainer and Zenject — is excluded from every `Generator~` check project. `TypeCheck` names the exclusion explicitly (`$(RepoRoot)/Runtime/Integrations/**/*.cs`), and `EditorCheck`, `TestCheck` and `EditorTestCheck` inherit it. So the shipped analyzers have **never** run over that tree.

That is not a theoretical gap. Four files carried the same rule-7 violation until #644's audit:

```csharp
AttributeMetadataCache cache = _metadataCache ?? AttributeMetadataCache.Instance;
if (cache == null) { return; }
```

`AttributeMetadataCache` is a `ScriptableObject`. `??` is a CLR-null test, so a **destroyed** injected instance is not replaced by the live `Instance` — it is passed through, and the very next line's Unity `==` then reads it as null and returns. VContainer's `RelationalSceneLoadListener` and Zenject's `RelationalComponentSceneInitializer` bail there: every relational field in that scene stays null, with no log. `WUH003` reports exactly this shape, and it never saw the file. The correct form was two lines away in `RelationalComponentAssigner` the whole time.

The four sites are fixed. This issue is the reason they could exist.

## Why the exclusion is there

Each integration asmdef is gated on a third-party DI package define, and none of the three ships a NuGet package a check project could reference. That is the same situation Odin was in, and #347 solved it the same way it needs solving here: a shim declaring only the base classes and interfaces the package's own sources alias.

## What would close this

- [ ] Enumerate the external surface each integration binds. Measured today: `Reflex.Core`, `Reflex.Extensions`, `Reflex.Injectors`, `Reflex.Enums`; `VContainer`, `VContainer.Unity`; `Zenject`. 19 files total, so the surface is small.
- [ ] Add shims under a new check project (or a fourth configuration of an existing one), declaring only what those files name — the Odin shim is the precedent for how narrow that can be.
- [ ] Load both analyzer DLLs, as every other check project does, so `WPROTO###` and `WUH###` both fire.
- [ ] Add the new project to `scripts/typecheck-controls.js`, with an anchor type from `Runtime/Integrations/**`, so the gate is proven able to fail over that tree rather than merely added.
- [ ] Extend `lint-typecheck-asmdef-references` and its `--probe` leg to the new project.

## Alternative, if the shims prove too costly

A source linter cannot answer this: the analyzer's own documentation records why the signal has to be the receiver's **type**, because `Vector2? p; p?.x` is correct and common. So the choice is the shims or an explicit, recorded decision that this tree is unanalyzed — in which case the decision belongs in `.llm/context.md` beside rule 7, so the next author knows the rule is unenforced exactly there.

Related: #644, #347, #598.

### LINKED ISSUE #688: A private relational field on a base class is silently never bound
Found by the #644 audit. Split out of that issue because it is the highest-impact finding in it and it deserves its own red-green.

## The defect

All three places that discover relational fields key on `component.GetType()` — the most derived type — and ask for fields without `BindingFlags.DeclaredOnly`:

- `Runtime/Core/Attributes/BaseRelationalComponentAttribute.cs` — the metadata path, `componentType.GetField(name, Instance | Public | NonPublic)`
- the same file's reflection fallback, `componentType.GetFields(Instance | Public | NonPublic)`
- `Editor/Tags/AttributeMetadataCacheGenerator.cs`, which builds the cache the first path reads

`Type.GetField`/`GetFields` without `DeclaredOnly` returns inherited **public, protected and internal** fields, and never an inherited **private** one. So this binds nothing:

```csharp
public abstract class Actor : MonoBehaviour
{
    [SiblingComponent]
    private Rigidbody2D _body;      // never discovered
}

public sealed class Enemy : Actor { }
```

`enemy.AssignRelationalComponents()` produces zero metadata for `_body`. It stays null, and because the field was never *discovered*, `LogMissingComponentError` never runs either — so there is no log, no warning, and no clue. The first symptom is a `NullReferenceException` somewhere else entirely.

Changing `private` to `protected` makes it work. That is what makes it a trap rather than a limitation.

## The package already believes the opposite

`RelationalComponentAssigner.ComputeHasRelationalAssignments` walks `current.BaseType` explicitly, and `HasRelationalAttributesViaReflection` does the same. So `HasRelationalAssignments(typeof(Enemy))` returns **true** — every DI integration then schedules and calls an assignment that binds nothing.

`docs/features/relational-components/relational-components.md` actively recommends `private` fields, and says nothing about inheritance.

## Why no test catches it

No test type in `Tests/` puts a relational attribute on a base class. `grep 'abstract class' Tests/**` returns only spatial, random and proto fixtures.

## What would close this

- [ ] A red test first: an `abstract` base declaring a `private` `[SiblingComponent]`, `[ChildComponent]` and `[ParentComponent]` field, and a `sealed` derived type. It must fail before the fix.
- [ ] Walk the type chain with `DeclaredOnly` at all three sites. A derived type may shadow a base field name with `new`, so collect per level and keep both `FieldInfo`s — each `SetValue` targets its own.
- [ ] The generator is the primary fix. Without it the cache has no entry for the base's private field, and the metadata path has nothing to look up however it searches.
- [ ] Decide what a shadowed name means and test it, rather than leaving it to whichever level the loop reaches first.
- [ ] Confirm the missing-component error now fires for a base-declared field that genuinely has no match, since today its silence is part of the defect.
- [ ] Update the relational-components guide to say inheritance is supported, now that it is.

## Watch for

This changes behaviour for existing consumers: a base-class private field that was silently unassigned starts being assigned, and one with no match starts logging an error it never logged. That is the fix, but it belongs in the CHANGELOG as a `Fixed` entry naming the shape.

Related: #644.

### LINKED ISSUE #689: Comparer-keyed pool caches root every comparer for the process lifetime
Found by the #643 audit. Split out because it is unbounded retention reachable from public API, and it has no owner in that issue's checklist.

## The defect

Four static caches in `Runtime/Utils/Buffers.cs` are keyed **by the comparer instance**:

- `SetBuffers<T>.HashSetCache` and `SortedSetCache`, filled by `GetHashSetPool` / `GetSortedSetPool`
- `DictionaryBuffer<TKey, TValue>.DictionaryCache` and `SortedDictionaryCache`, filled by `GetDictionaryPool` / `GetSortedDictionaryPool`

Each is a `static readonly ConcurrentDictionary<IComparer<T>, WallstopGenericPool<…>>` on a generic type, so it lives for the process. `GetOrAdd` interns the caller's comparer as the **key** — a strong reference — with no eviction, no weak key, no size bound and no purge hook.

`GlobalPoolRegistry` holds pools by `WeakReference` precisely so they can be collected. These caches hold them **strongly**, so neither the pool nor the comparer can ever be collected.

The only removal path is `DestroyHashSetPool` / `DestroySortedSetPool` / `DestroyDictionaryPool` / `DestroySortedDictionaryPool`. Those have **zero production call sites** — the only callers are `Tests/Runtime/Utils/BuffersTests.cs`.

## How a consumer reaches it

Every one of these takes a caller-supplied comparer through public API:

- `new SpatialHash2D<T>(cellSize, comparer)` → `GetHashSetPool(_comparer)` on every `distinct: true` query. Same for `SpatialHash3D`.
- `SerializableDictionary` → `GetHashSetPool(_dictionary.Comparer)`
- `SerializableHashSet` → `GetHashSetPool(SetComparer)`, overridden to `Set.Comparer`
- `SerializableSortedDictionary`

So `new SpatialHash2D<Enemy>(4f, new ProximityComparer(levelRoot))` per level roots that comparer, and the pool behind it, forever. If the comparer is a `MonoBehaviour` or `ScriptableObject`, or a closure capturing one — the ordinary shape for a Unity comparer — that Unity object survives scene unload, and every play session when Domain Reload is off. Cardinality is whatever the game's comparer-instance cardinality is; nothing in the package bounds it.

## What would close this

- [ ] A red test first: rent through N distinct comparers, drop every reference to them, force a collection, and assert the cache does not still hold them. A `WeakReference` to each comparer is the observable.
- [ ] Decide the model. The obvious candidates, in rough order of how much they change:
  - key by comparer **identity through a weak key**, with entries evicted when the comparer dies;
  - bound the cache and evict least-recently-used, since a pool is a cache and losing one costs an allocation rather than correctness;
  - key by comparer **type** rather than instance where the comparer is stateless, which collapses the common case to one entry and leaves stateful comparers as the only growth;
  - or refuse to cache a comparer the package did not create, and hand those callers a fresh pool they own.
- [ ] Whichever it is, give `Destroy*Pool` a production caller or delete it — a removal path only tests use is not a removal path.
- [ ] Assert the bound after an adversarial run of unique comparers, and after quiescence, per #643's own retention section.

## Related, same file, smaller

`PartitionPooledEnumerator` double-wraps every partition (`IEnumerableExtensions.cs`), leaking one `DisposalLease` slot per partition when the consumer does not dispose each `Current` — which the documented contract permits, because the enumerator's own `Dispose` is the safety net, and that net claims only the first lease.

Related: #643, #632.

### LINKED ISSUE #690: security: keep the Unity license cache out of the artifacts tree
Found while wiring credential redaction in #685. Not a live leak today, but it is one configuration change away from being one, and it contradicts an invariant this repository states elsewhere.

## What happens

`scripts/unity/export-unitypackage.sh:254` overrides `UNITY_TEST_PROJECT_DIR` into the artifacts tree, and `scripts/unity/run-unity-docker.sh:77` then places `UNITY_LICENSE_CACHE_DIR` inside it:

```
.artifacts/unity/unitypackage-project/.unity-license-cache/
.artifacts/unity/unitypackage-smoke-project/.unity-license-cache/
```

`run-unity-docker.sh:718-719` bind-mounts that cache onto `/root/.local/share/unity3d` and `/root/.config/unity3d`. `docker run` passes no `--user`, so the container is root and the licensing client writes the cache root-owned.

The new redaction step found at least one file in there, `config-unity3d/Unity/Unity.Entitlements.Audit.log`, that matches a declared credential pattern. It could not rewrite the file because of the root ownership, and correctly failed closed rather than passing silently. `retry-license.sh:43-46` and `validate-license-setup.sh:130-133` both name `Unity_lic.ulf` and `UnityEntitlementLicense.xml` in those same two directories, and those carry license identity by definition, so the audit log is likely the first of several rather than the only one.

## Why it is not a live leak today

Four independent reasons, all verified against the real artifact of run 33597641377:

- The two Docker export jobs upload four explicit paths, and `.unity-license-cache` is a sibling of all of them rather than under any.
- The uploaded `unitypackage-smoke-diagnostics` artifact holds 22 files, none from the cache.
- No workflow sets `include-hidden-files: true`, and `actions/upload-artifact` excludes dot-prefixed paths by default.
- No upload names a `unitypackage-*-project` directory wholesale.

`.gitignore:416` also covers `.unity-license-cache/`, so there is no commit risk.

The two nearest misses are widening a diagnostics upload to the project directory, or adding `include-hidden-files: true` to one of these jobs.

## Why it should still change

`run-ci-tests.ps1:2150-2151`, inside the `# SECURITY:` block at 2148-2154, states the rule for the Windows path: the activation log lives under a non-uploaded temp directory, never under the artifacts path. The Docker path has no equivalent rule. `run-unity-docker.sh:76` explains the cache sits in the project volume to avoid workspace ownership and commit risks, which solves the git problem by moving the material into the CI upload tree instead.

Depending on a default that excludes dot directories is not the same as keeping credential material out of the tree.

## Suggested fix

Point `UNITY_LICENSE_CACHE_DIR` at `RUNNER_TEMP` (or the system temp directory outside the workspace) for the CI path, matching what the Windows runner already does. Keep the current in-project default for the local devcontainer case if that is what avoids the ownership problem there.

Add a contract assertion that no path under an uploaded artifacts directory is a license cache, so the rule holds rather than relying on an upload default.

## Also worth confirming

What kind of credential material `Unity.Entitlements.Audit.log` actually holds is unconfirmed. What is proven is only that it matched at least one of the seven declared patterns, because redaction attempts a write solely when a pattern matches. Reading it needs root inside the container. Confirm the shape before deciding whether the existing files also need scrubbing rather than only relocating.

One behavioural note for whoever picks this up: the redactor throws on the first unwritable hit, and `Unity.Entitlements.Audit.log` sorts before its neighbours, so nothing after it was examined.

### PR DESCRIPTION
**Pull request by wallstop**
**Why:** A relational field declared `private` on a base class was never bound and never logged, and four static pool caches kept every comparer a game handed them — and any scene object one captured — alive for the process.

**What:**

- Field discovery walks the type chain with `DeclaredOnly`, swept to all thirteen sites; a base-declared field also stops falling back to `FieldInfo.SetValue` on every assignment.
- The comparer-keyed set and dictionary pool caches are bounded and evict least-recently-used; `Buffers.ComparerPoolMaxDistinctEntries` is the knob.
- Five pooled leases survive a throw, one is no longer lent to a live `DropdownField`, and `Partition` stops burning a lease slot per partition.
- A poisoned `ScriptableObjectSingleton` recovers, `Instance` never serves a destroyed asset, and `[AutoLoadSingleton]` re-arms each play session.
- The Unity license cache moves out of the uploaded artifacts tree, the license-year walk stops reading gitattributes it cannot use, and `Runtime/Integrations` is compiled and analyzed for the first time.

Fixes #688
Fixes #689
Fixes #690
Fixes #680
Fixes #687

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches core relational binding, static pool caches, and singleton lifecycle used broadly at runtime; incorrect field discovery or pool eviction could cause subtle assignment or memory/aliasing bugs, though the change set is heavily tested.
> 
> **Overview**
> Fixes **relational component binding** so `[SiblingComponent]`, `[ChildComponent]`, and `[ParentComponent]` on **private base-class fields** are discovered and assigned. Discovery now walks the inheritance chain (`GetInstanceFieldsIncludingBaseTypes`), cache-driven metadata is matched against live fields (not name-only), and field accessors use the **declaring type** so assignments stay on the fast path.
> 
> **Pooling and editor UI:** comparer-keyed set/dictionary pools are **LRU-bounded** via new `Buffers.ComparerPoolMaxDistinctEntries` and `ComparerKeyedPoolCache`; `Partition` no longer double-leases lists; pooled dropdown choices are rewired on attach/detach; several paths add **try/finally** so throws do not drop leases. **`TextureScale`** waits for worker threads before returning pooled color buffers.
> 
> **Singletons:** `ScriptableObjectSingleton` logs load failures and returns null instead of a poisoned `Lazy`, reloads after a destroyed asset (without re-searching when the asset is genuinely absent), and resets metadata latches on clear/reload. **`[AutoLoadSingleton]`** clears its per-session execution set when Enter Play Mode skips domain reload.
> 
> **Tooling:** adds **`IntegrationCheck`** as a fifth `dotnet` gate over `Runtime/Integrations/**` with Reflex/VContainer/Zenject shims; docs/CHANGELOG note five typecheck controls. Workflow comments document Unity artifact redaction scope after the Docker license cache move (#690).
> 
> Tests cover inheritance edge cases, singleton recovery, and an `AllocationProbe` for GC measurement.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 14e80b9be22b0ea6d7ae04635597c6e815233c2b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: Runtime/Core/Helper/ReflectionHelpers.cs [hunk 1]
```diff
@@ -381,6 +399,128 @@ internal static FieldInfo[] GetFieldsCached(Type type, BindingFlags flags)
...
+            Type current = type;
+            while (current != null && current != typeof(object))
+            {
+                foreach (FieldInfo field in GetFieldsCached(current, DeclaredInstanceFieldsFlags))
+                {
+                    if (seenNames.Add(field.Name))
+                    {
+                        collected.Add(field);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (wallstop) [reply Runtime/Core/Helper/ReflectionHelpers.cs:520]: Correct, and thank you — this is the defect the PR exists to fix, reintroduced one level narrower. Fixed in `cd884128`. **Fix + sweep + rule**, in that order.

**Fix.** The walk collapses a name only when the base declaration is **visible**. A private field is invisible to a derived type, so a same-named field there is a second distinct field and both are kept; a `new`-hidden `protected`/`public`/`internal` field is genuine name hiding and still collapses to the most derived one.

Measured in the editor `6000.4.6f1` rather than argued:

```text
distinct private pair       -> 2 declarations: RelationalPrivateNameDerived RelationalPrivateNameBase
new-hidden protected pair   -> 1 declaration:  RelationalShadowDerived
```

**Sweep — and fixing the walk alone was not enough.** The other location you flagged, `BaseRelationalComponentAttribute`, resolved every cached metadata entry by **name**, so two same-named fields would have bound the derived one twice and left the base one unbound, silently. The generator records one entry per declaration in the same chain order the walk produces, so entries now resolve by **position within a name**. `GetFieldsWithAttribute` needed nothing: it already reports every declaration, which is what a caller asking which fields carry an attribute wants.

**Rule.** `RelationalPrivateNameBase`/`Derived` are new fixtures asserting both fields bind, beside the existing `RelationalShadowBase`/`Derived` pair asserting the hiding case still collapses — so the two halves of the decision are pinned against each other. 137 assertions across the nine relational, `[WNotNull]` and `[ValidateAssignment]` fixtures pass in the editor, zero failures.

One aside worth recording: the first draft of the occurrence counter used an unguarded `TryGetValue` out-read, and this package's own `WUH008` failed the build over it. The analyzer working on its own author.

File: Runtime/Core/Attributes/BaseRelationalComponentAttribute.cs [hunk 2]
```diff
@@ -230,16 +243,39 @@ out AttributeMetadataCache.RelationalFieldMetadata[] cachedFields
...
+                    )
+                        ? previousOccurrence
+                        : 0;
+                    occurrences[cachedField.fieldName] = occurrence + 1;
+
+                    FieldInfo field = ReflectionHelpers.GetInstanceFieldIncludingBaseTypes(
+                        componentType,
                         cachedField.fieldName,
-                        BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic
+                        occurrence
```
[the remark this answers is not in the document]
Comment (wallstop) [reply Runtime/Core/Attributes/BaseRelationalComponentAttribute.cs]: Right on both, and they are the same mistake twice: the previous commit tried to reconstruct **which declaration** a cache entry meant from a format that is keyed by name and cannot say. Fixed in `ed30f9a5`.

**Fix.** The cached path is driven by the **live fields** now, not by the cache entries. The attribute is on the field, so nothing has to be inferred; the cache only has to say which *names* are relational for this kind, which is exactly what a name-keyed format can say. Position never enters it, so the failure you named — a derived `_collider` with no attribute, or with `[ChildComponent]` where the base has `[SiblingComponent]`, taking index 0 and failing the attribute check — cannot happen.

**Sweep — your second finding, the element type, is the same root cause and is fixed with it.** `TryGetElementType` is still keyed by component type and name, so two same-named fields of different types do share one slot. `ChooseElementType` keeps the recorded type except where it cannot be describing this field: if it is neither the live element type nor related to it by assignability, it is the *other* field's and the live one wins. The cache keeps its role in the ordinary case and stops steering the search at a wrong component type in the pathological one.

**Rule.** Two fixtures cover exactly what the two findings named — `RelationalPrivateNameDerived` (both fields carry the attribute, both bind) and `RelationalPrivateNameUnrelatedDerived` (the derived field reuses the name with **no** relational attribute, and the base field still binds). The `new`-hidden `RelationalShadow*` pair sits beside them asserting genuine name hiding still collapses to the most derived declaration, so the three cases pin each other.

145 assertions across the eleven relational, `[WNotNull]` and `[ValidateAssignment]` fixtures pass in the editor `6000.4.6f1`, zero failures; `npm run typecheck:unity` green across all 19 builds.

The dead `occurrence` overload the previous commit added is gone rather than left unused, and `GetInstanceFieldIncludingBaseTypes` now says in its own remarks that a name is not an identity here — a caller that must reach both declarations enumerates instead of asking by name.

[the remark this answers is not in the document]
Comment (wallstop) [reply Runtime/Core/Attributes/BaseRelationalComponentAttribute.cs]: Correct, and fixed in `ed30f9a5` alongside the occurrence finding — they share a root cause, so recording the answer on this thread too.

`TryGetElementType` is still keyed by component type and field name, and I am not changing the serialized cache format for a pathological case. Instead `ChooseElementType` refuses a recorded type that cannot be describing this field: if it is neither the live element type nor related to it by assignability in either direction, it is the *other* same-named field's, and the live one wins. The cache keeps its role in the ordinary case — where the two agree — and stops steering the search at a wrong component type in the case you named.

The live element type is always available, because the cached path is now driven by the live fields rather than by cache entries, so `GetFieldKind(field.FieldType, out Type actualElementType)` has already run by the time the choice is made.

Verified green: `npm run typecheck:unity` across all 19 builds, and Unity Tests passing on all four editors in both editmode and playmode.
